### ISSUE TITLE
Handling events with MediaPlayerEventAdapter in JavaFX 

### ISSUE THREAD

**Issue by courteous** [caprica/vlcj#1082]
Hello Mark, 

After struggling for couple of days, I have prepared a very simple maven project of a vlcj + javafx . 

[https://github.com/courteous/vlcjTestMute](https://github.com/courteous/vlcjTestMute)

Basically in this project consist of 2 files the controller "FXMLDocumentController.java"  and the  "MediaPlayerEventAdapterHelper.java" . The GUI has 2 buttons i.e. "play" button and "mute" button.  The lines 

```
		System.out.println("embeddedMediaPlayer mute event called ");
		mediaPlayerObj.audio().setMute(muted);
```

but the media player is not muted. Do you have an idea what could be the reason of this? Why is the embeddedMediaPlayer not muting in this case?

I am able to reproduce the same behavior  from ticket #1079 (i.e. the sound is not being muted) in a simple project. 

PS: in order to run the test the only thing that is required is to change the line in the playButtonAction.
String file = "file:///C:/Users/courteous/Downloads/sound.wav";
to some local sound.wav file

**Comment by caprica** [comment]
Your code is not right I'm afraid.

You are dealing with events in a very strange way, I sincerely mean no offence, but I have no idea how you arrived at that solution for events.

Get rid of *everything* you've done with media player events, and change your listener thusly:
```
	@FXML
	private void muteButtonAction(ActionEvent event) {
		try {
			embeddedMediaPlayer.audio().mute();              // <--- doesn't get any simpler than this
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

```
100% working.

How simple is that.

An aside...

I appreciate your code is just an example, but don't ordinarily create your media player in the playButtonAction method - create the media player once up-front in your constructor.

**Comment by caprica** [comment]
I will add there are a lot of examples provided with vlcj that show how to use events properly, please have a look at the examples - I cover just about everything you can in vlcj with those examples.

**Comment by courteous** [comment]
Hello Mark , 

thank you for the feedback and your support. I was able to find where the problem was and i wanted to share it since i find it very tricky to find and i hope this will help someone else in the future. I added an example of a non working events to  git git respiratory above. Basically here is the code where the events are NOT working when playing my local camera.
```
        MediaApi mediaApi =  this.embeddedMediaPlayer.media();
        String sourceOfMedia = "dshow://" ;
	    
        String cameraDevice = "Logitech Webcam C930e";
	String microfonDevice = "Mikrofon (2- Logitech Webcam C930e)";
	   	   
	String[] options = { 	
	    						
	    						":dshow-vdev=" + cameraDevice,  
	    						":dshow-adev=" + microfonDevice, 
	    						
	    						":sout=#duplicate{"
    							+ "dst=display"
	    						+ "}' "
	    						
	    						};
	  	    
	Boolean prepared = mediaApi.prepare(sourceOfMedia);
	System.out.println("Is prepared " + prepared);

	Boolean isPlaying = mediaApi.play(sourceOfMedia, options);
```

Now at first there is nothing fancy about this code i.e. just a normal way to play the camera. However if you test it you will notice that no mute/unmute is possible. This drove me mad for quite some time.  
Now here is the trick i.e. a missing white-space before 
":sout=#duplicate{"
i.e. change that to 
" :sout=#duplicate{"

If you put that missing white-space there all events magically start work as they are supposed to per design.   The funny thing is that this kind behaviors is really really really difficult to spot, at least for me that was the case.  The reason for that is because the method duplicate is indeed working correctly since i am able to see the video of the camera locally on my laptop. 
I do not know if the problem is in libvlc or in vlcj of parsing those string options, but after testing it i am sure that the whitespace have some meaning to the way how events are handled.  Maybe something can be added to the parser to spit at error in such cases. 

Hope this helps someone else.

**Comment by caprica** [comment]
Sadly there is no API in LibVLC for that "sout" functionality, so this kind of problem with mis-typed strings is not impossible. Those sout strings are passed directly to LibVLC.

You mention about maybe adding something to the (native) parser to detect that as an error case - it might not be so obvious because that sout syntax can be deceptively "rich", in the sense that the presence or absence of the # might actually be significant and trigger different intended behaviours.

**Comment by caprica** [comment]
An unrelated point... doing prepare before play like you have done is redundant. If you use prepare, then you should just invoke play on the controls API, not play on the media API. In your case it is effectively preparing the media twice, once in your explicit prepare, and once again when you invoke play on the media API. Generally speaking, prepare is rarely required.

**Comment by courteous** [comment]
Hello Mark,
i just updated the repository above i.e. commit [`ec5352a`.](https://github.com/courteous/vlcjTestMute/commit/ec5352afaf1d5eafe7cd7e3e3ab9b992558fdff3) 
In that example i am using the ":sout=#transcode" and the method :duplicate from the vlclib.  
If i remove the whitespace in question the events do not work, if i add the whitespace the "transcoding " does not work.  
Do you know if the string that i am using is correct.  i.e.


	    String[] options = { 	
	    						
	    						":dshow-vdev=" + cameraDevice,  
	    						":dshow-adev=" + microfonDevice, 
	    						" :sout=#transcode{vcodec=h264,fps=24vb=750,scale=Auto,width=640,height=360,acodec=mp4a,ab=128,channels=2,samplerate=44100,threads=4,scodec=none}"
	    						// start duplicating the streams i.e. send it over the network and display it locally 
	    						+ ":duplicate{"
	    						+ "dst=standard{access=udp,mux=ts,"
	    						+ "dst=172.26.15.107:5555"
	    						+ "},"
	    						+ "dst=display"
	    						+ "} "
	    						
	    						};

Here i am creating an array called options, notice there are no "+" at the end of cameraDevice and microfonDevice. 
Is there an alternative way to implement those 2 functions transcoding and duplicate without using ":sout" trough vlcj? Since for some reason this is definitely does not work. If i remove the whitespace the transcoding is working but events stop working. 
Any idea is really appreciated.

**Comment by courteous** [comment]
I have removed the custome camera and left only the  default build in camera in the latop with the following string 

	    						```
":sout=#transcode{vcodec=h264,fps=24vb=750,scale=Auto,width=640,height=360,acodec=mp4a,ab=128,channels=2,samplerate=44100,threads=4,scodec=none}"
	    						// start duplicating the streams i.e. send it over the network and display it locally 
	    						+ ":duplicate{"
	    						+ "dst=standard{access=udp,mux=ts,dst=172.26.15.107:5555},"
	    						+ "dst=display"
	    						+ "}"
```

Still no luck, if there is no whitespace before ":sout" then transcoding is working but events are not.

**Comment by courteous** [comment]
Hello Mark, 
after debbugging for a while here is what i found 
1) When transcoding is working but evenets are not in the moment we issue 
audioApi.setMute(true);

the libvlc is beeing called and the method 
libvlc_audio_set_mute(mediaPlayerInstance, mute ? 1 : 0); 
is beeing executed.  this is the non working example 

![grafik](https://user-images.githubusercontent.com/1891541/126781605-f886ae59-df07-48d2-b5c2-73b91aa20015.png)

2) when the whitespace is there the events are working but  the transcoding is not. in such case however we see a difference i.e.
there is a callback from libvlc back i.e. 
a) we execute  "audioApi.setMute(true);"
b) then we see the call to  "libvlc_audio_set_mute(mediaPlayerInstance, mute ? 1 : 0);"
d) then we get a callback from the libvlc i.e. see  the screenshot below 
e) then the overriten method muted in the MediaPlayerEventAdapter is being executed. 

![grafik](https://user-images.githubusercontent.com/1891541/126781791-4783433b-1511-432f-bf7e-ca8c63728302.png)



So basically in the first case when trascoding is working we do not get that callback.
